File: Python/Easy/validParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        
        for char in s:
            if char in '({[':
                stack.append(char)
            elif char in ')}]':
                if not stack or stack[-1] != mapping[char]:
                    return False
                stack.pop()
                
        return not stack
    
        
def main():
    case1 = "()" 
    case2 = "()[]{}"
    case3 = "(]"
    case4 = "([])"
    case5 = "([)]"

    solution = Solution()
    print(solution.isValid(case1))
    print(solution.isValid(case2))
    print(solution.isValid(case3))
    print(solution.isValid(case4))
    print(solution.isValid(case5))
    
if __name__ == "__main__":
    main()
        

File: JS/Easy/validParentheses.js
let isValid = function(s) {
    let stack = []
    const mapping = new Map();
    mapping.set(')', '(');
    mapping.set('}', '{');
    mapping.set(']', '[');

    for (char of s) {
        if ("({[".includes(char))
            stack.push(char);
        else if (")}]".includes(char)) {
            if (stack.length !== 0 && stack[stack.length - 1] !== mapping.get(char))
                return false;
            stack.pop();
        }
    }

    return stack.length === 0;
}

const s1 = "()";
console.log(isValid(s1));

const s2 = "()[]{}";
console.log(isValid(s2));

const s3 = "(]";
console.log(isValid(s3));

const s4 = "([])";
console.log(isValid(s4));

const s5 = "([)]";
console.log(isValid(s5));